src/passes/TypeSSA.cpp

```diff
@@ -147,14 +148,136 @@ std::vector<HeapType> ensureTypesAreInNewRecGroup(RecGroup recGroup,
+  }
+  void visitArrayNew(ArrayNew* curr) {
+    news.push_back(curr);
+    visitExpression(curr);
+  }
+  void visitArrayNewData(ArrayNewData* curr) {
+    news.push_back(curr);
+    visitExpression(curr);
+  }
+  void visitArrayNewElem(ArrayNewElem* curr) {
+    news.push_back(curr);
+    visitExpression(curr);
+  }
+  void visitArrayNewFixed(ArrayNewFixed* curr) {
+    news.push_back(curr);
+    visitExpression(curr);
+  }
+
+  // Find casts to exact types. Allocations of these types will not be able to
+  // be optimized.
+  template<typename Cast> void visitCast(Cast* cast) {
+    if (auto type = cast->getCastType(); type.isExact()) {
+      exactTypes.insert(type.getHeapType());
+    }
+  }
+  void visitRefTest(RefTest* curr) { visitCast(curr); }
+  void visitRefCast(RefCast* curr) { visitCast(curr); }
+  void visitBrOn(BrOn* curr) {
+    if (curr->op == BrOnCast || curr->op == BrOnCastFail) {
+      visitCast(curr);
+    }
+  }
+
+  void visitExpression(Expression* curr) {
+    // Look at the constraints on this expression's operands to see if it
+    // requires an exact operand. If it does, we cannot optimize allocations of
+    // that type. As an optimization, do not let control flow structures or
+    // branches inhibit optimization since they can safely be refinalized to use
+    // new types as long as no other instruction expected the original exact
+    // type. Also allow optimizing if the instruction that would inhibit
+    // optimizing will not be written in the final output.
+    if (Properties::isControlFlowStructure(curr) ||
+        Properties::isBranch(curr) ||
+        Properties::hasUnwritableTypeImmediate(curr)) {
+      return;
+    }
+
+    // Also do not let unreachable instructions inhibit optimization, as long as
+    // they are unreachable because of an unreachable child. (Some other
+    // unreachable instructions, such as a return_call, can still require an
+    // exact operand and may inhibit optimization.)
+    if (curr->type == Type::unreachable) {
+      for (auto* child : ChildIterator(curr)) {
+        if (child->type == Type::unreachable) {
+          return;
+        }
+      }
+    }
+
+    struct ExactChildTyper : ChildTyper<ExactChildTyper> {
+      Analyzer& parent;
+      ExactChildTyper(Analyzer& parent)
+        : ChildTyper(*parent.getModule(), parent.getFunction()),
+          parent(parent) {}
+
+      void noteSubtype(Expression**, Type type) {
+        for (Type t : type) {
+          if (t.isExact()) {
+            parent.exactTypes.insert(t.getHeapType());
+          }
+        }
+      }
+
+      // Other constraints do not matter to us.
+      void noteAnyType(Expression**) {}
+      void noteAnyReferenceType(Expression**) {}
+      void noteAnyTupleType(Expression**, size_t) {}
+      void noteAnyI8ArrayReferenceType(Expression**) {}
+      void noteAnyI16ArrayReferenceType(Expression**) {}
+
+      Type getLabelType(Name label) { WASM_UNREACHABLE("unexpected branch"); }
```

Comment:
Isn't this mandatory to implement?

Reply:
Yes, you get a compiler error if you don't implement it. But this implementation is sufficient because we never use this class to analyze a branch.